oml/postprocessors/pairwise_embeddings.py

```diff
@@ -0,0 +1,43 @@
+import torch
+from torch import Tensor
+
+from oml.datasets.vectors_pairs import pairwise_inference
+from oml.interfaces.models import IPairwiseDistanceModel
+from oml.interfaces.postprocessor import IPostprocessor
+from oml.utils.misc_torch import assign_2d
+
+
+class PairwiseEmbeddingsPostprocessor(IPostprocessor):
+    def __init__(self, pairwise_model: IPairwiseDistanceModel, top_n: int):
+        assert top_n > 0
+
+        self.model = pairwise_model
+        self.top_n = top_n
+
+    def process(self, distances: Tensor, emb_query: Tensor, emb_gallery: Tensor) -> Tensor:  # type: ignore
+        n_queries = len(emb_query)
+        n_galleries = len(emb_gallery)
+
+        assert list(distances.shape) == [n_queries, n_galleries]
+
+        # pick top n galleries for each query with the smallest distances
+        top_n = min(self.top_n, n_galleries)
+        ii_top = torch.topk(distances, k=top_n, largest=False)[1].view(-1)
+
+        # create (n_queries * top_n) pairs of each query and related galleries and re-estimate distances for them
+        emb_query = emb_query.repeat_interleave(top_n, dim=0)
+        emb_gallery = emb_gallery[ii_top]
+        distances_upd = pairwise_inference(self.model, emb_query, emb_gallery)
+        distances_upd = distances_upd.view(n_queries, top_n)
+
+        # update distances for top-n galleries, keeping the order of rest of the galleries (we use offset for it)
+        offset = distances_upd.min(dim=1)[0] - distances.min(dim=1)[0] + torch.finfo(torch.float32).eps
+        distances += offset.unsqueeze(-1)
+        distances = assign_2d(x=distances, indeces=ii_top.view(n_queries, top_n), new_values=distances_upd)
```

Comment:
I'm not sure that it is a good idea to keep somehow old distances. It could be a source of very subtle errors, like when `top_n` here is less than say `cmc_top_k` or when a user just forgot that the postprocessor is turned on whereas he thinks that it is turned off. I'd suggest to spoil all the distances such that it will be clear that distance matrix has been changed somehow, and it will be a hint for a user to check what could have changed the distance matrix. I'd suggest to fill the rest of the distance matrix with `inf`.

Reply:
And I don't know where it is better to add that we need to warn (or better to raise an error, because who reads warnings?) user that the `top_n` is less than `top_k` values for `cmc`, `precision` etc.

Reply:
discussed offline